text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,278 @@
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. The back-end for CDK Construct Hub starts with a custom-built event source,
+   which implements the logic necessary to detect new packages that are relevant
+   to the CDK Construct Hub, and sends messages to an SQS queue for further
+   processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
+   `time`      | The timestamp at which the version was created
+   `integrity` | An integirity check for the complete record
+
+1. A Lambda function then picks up messages from the SQS queue and prepares the
+   artifacts consumed by the front-end application, stored in a dedicated S3
+   bucket using the following key format:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly.json`
+
+   This function validates the contents of the `assembly` to ensure the message
+   was well-formed. In case the message is found to be incoherent, it is sent to
+   a dead-letter queue and an alarm is triggered.
+
+   It also creates or updates the `assemblies/${assembly.name}/versions.json`
+   object, which contains a list of all versions ever indexed for the package,
+   matched to the versions' current status (`deprecated`, `latest`, ...):
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `v${major}` | All versions part of this major version stream, and their status
+
+   [SemVer]: https://semver.org/spec/v2.0.0.html
+
+1. A series of Lambda functions prepare language-specific assembly files for
+   each configured language, with adjusted naming conventions, and updated
+   sample code fragments, and proceeds to store those at:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly-${lang}.json`.
+   This transformation is backed by the `jsii-rosetta` tool, which is part of
+   the [jsii project][jsii].
+
+   [jsii]: https://aws.github.io/jsii
+
+1. A Lambda function keeps the `latest.json` object updated with the last 20
+   package versions indexed (according to the `time` field reported by the
+   ingestion component).
+
+1. A Lambda function keeps the `catalog.json` object updated with the latest
```

Comment:
How are we planning to handle concurrency? Are these updates serialized through a queue?